### PR TITLE
MGMT-19523: Fix validation looping in ODF

### PR DESCRIPTION
**Pull request by danmanor**
### The assisted-service supports two deployment modes for OpenShift Data Foundation:

- `compact` - A deployment with at least 3 control plane nodes and no workers. In this deployment the ODF workloads will run on the control plane nodes.
- `standard` - A deployment with at least 3 control plane nodes and at least 3 worker nodes. In this deployment ODF workloads will run on the worker nodes.

Prior to [this pull request](https://github.com/openshift/assisted-service/pull/7036), compact mode was restricted to exactly 3 control plane nodes. This meant determining the number of workers was straightforward—simply count the hosts beyond the 3 control plane nodes. However, with the introduction of 4- and 5-node control plane configurations, this assumption no longer held. To determine the number of control plane (master) and worker nodes, it became necessary to check the assigned roles of all cluster hosts.

This role information is crucial for determining the deployment mode, which in turn dictates whether to validate ODF requirements on the control plane or worker nodes.

### Host Role Assignment Mechanism
Host roles are automatically assigned at regular intervals as follows:

Count the current number of control plane (master) nodes in the cluster.
If additional master nodes are required, assign the current host as master and run all validations for this role.
If no additional master nodes are needed, assign the host as a worker.

### Issue with Role Assignment and ODF Validations

Part of the host validation process (step 2) includes ODF-specific checks, which depend on the deployment mode. These checks rely on the accurate count of master and worker nodes in the cluster.

However, an issue arose because the role assignment in step 2 updated the host's role for validations in memory but did not persist the updated role to the cluster's internal data structure. As a result, ODF validations—requiring an accurate view of all host roles—produced inconsistent results due to the mismatch between the in-memory role assignment and the cluster's stored host roles.

### Solution

This PR updates the matching host in the cluster struct, adds some logs and changes some validation messages

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [x] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [x] Cloud
- [ ] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [x] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/host/monitor.go [hunk 1]
```diff
@@ -162,33 +162,45 @@ func (m *Manager) clusterHostMonitoring() int64 {
 			break
 		}
 
+		log.Debug("Started cluster host monitoring cycle")
```
Comment (paul-maidment) [inline internal/host/monitor.go:165]: Will these debug messages be left in the code?

File: internal/operators/manager_test.go [hunk 2]
```diff
@@ -479,7 +479,7 @@ var _ = Describe("Operators manager", func() {
 			Expect(results).To(ContainElements(
 				api.ValidationResult{Status: api.Success, ValidationId: string(models.ClusterValidationIDLsoRequirementsSatisfied), Reasons: []string{}},
 				api.ValidationResult{Status: api.Failure, ValidationId: string(models.ClusterValidationIDOdfRequirementsSatisfied),
-					Reasons: []string{"A cluster with only masters or with a minimum of 3 workers is required."}},
+					Reasons: []string{"The cluster must either have no dedicated worker nodes or at least three. Add or remove hosts, or change their roles configurations to meet the requirement."}},
```
Comment (paul-maidment) [inline internal/operators/manager_test.go:482]: NIT:
Should read
"The cluster must either have no dedicated worker nodes or at least three. Add or remove hosts, or change their role configurations to meet the requirement."

File: internal/operators/odf/validations.go [hunk 3]
```diff
@@ -18,11 +18,17 @@ type odfClusterResourcesInfo struct {
...
+	if cluster.ID != nil {
+		log = log.WithField("cluster", cluster.ID.String())
+	}
 
 	mode := getODFDeploymentMode(cluster, o.config.ODFNumMinimumHosts)
+	log.Debugf("ODF validate cluster - mode: %s", string(mode))
 
 	if mode == unknown {
-		status = "A cluster with only masters or with a minimum of 3 workers is required."
+		status = "The cluster must either have no dedicated worker nodes or at least three. Add or remove hosts, or change their roles configurations to meet the requirement."
```
Comment (paul-maidment) [inline internal/operators/odf/validations.go:31]: NIT:
Change to
"The cluster must either have no dedicated worker nodes or at least three. Add or remove hosts, or change their role configurations to meet the requirement."

File: internal/operators/manager_test.go (hunk 2 above)
Comment (danmanor) [reply internal/operators/manager_test.go:482]: I agree with you, but I wanted to use the same as in - https://github.com/openshift/assisted-service/blob/0e3133ea2cef252e63573178db88186e84b927e1/internal/cluster/validator.go#L362 (I didn't wrote that one, AFAIU it is UX.) Lets keep compatibility

File: internal/operators/odf/validations.go (hunk 3 above)
Comment (danmanor) [reply internal/operators/odf/validations.go:31]: I agree with you, but I wanted to use the same as in - https://github.com/openshift/assisted-service/blob/0e3133ea2cef252e63573178db88186e84b927e1/internal/cluster/validator.go#L362 (I didn't wrote that one, AFAIU it is UX.) Lets keep compatibility

File: internal/host/monitor.go (hunk 1 above)
Comment (danmanor) [reply internal/host/monitor.go:165]: Yes

**Review by paul-maidment — APPROVED** [review]
/lgtm

**Pull request merged**